dojo/pipeline.py

```diff
@@ -53,3 +59,43 @@ def social_uid(backend, details, response, *args, **kwargs):
 def modify_permissions(backend, uid, user=None, social=None, *args, **kwargs):
     if kwargs.get('is_new'):
         user.is_staff = False
+        if settings.GITLAB_PROJECT_AUTO_IMPORT is True:
+            # Add engagement creation permission if auto_import  is set
+            user.user_permissions.set([Permission.objects.get(codename='add_engagement', content_type=ContentType.objects.get_for_model(Engagement)), Permission.objects.get(codename='add_test', content_type=ContentType.objects.get_for_model(Test)), Permission.objects.get(codename='change_test', content_type=ContentType.objects.get_for_model(Test))])
+
+
+def update_product_access(backend, uid, user=None, social=None, *args, **kwargs):
+    if settings.GITLAB_PROJECT_AUTO_IMPORT is True:
+        # Get all product names
+        prod_names = [prod.name for prod in Product.objects.all()]
+        # Get Gitlab access token
+        soc = user.social_auth.get()
+        token = soc.extra_data['access_token']
+        # Get user's projects list on Gitlab
+        gl = gitlab.Gitlab(settings.SOCIAL_AUTH_GITLAB_API_URL, oauth_token=token)
+        # Get each project path_with_namespace as future product name
+        projects = gl.projects.list(membership=True, all=True)
+        project_names = [project.path_with_namespace for project in projects]
+        # Create product_type if necessary
+        try:
+            product_type = Product_Type.objects.get(name='Gitlab Import')
+        except ObjectDoesNotExist:
+            product_type = Product_Type.objects.create(name='Gitlab Import')
+        # For each project: create a new product or update product's authorized_users
+        for project_name in project_names:
+            if project_name not in prod_names:
+                # Create new product
+                product = Product.objects.create(name=project_name, prod_type=product_type)
+                product.authorized_users.add(user)
+                product.save()
+            else:
+                # Update product
+                product = Product.objects.get(name=project_name)
+                product.authorized_users.add(user)
+                product.save()
+        # For each product: if user is not project member any more, remove him from product's authorized users
+        for product_name in prod_names:
+            if product_name not in project_names:
+                product = Product.objects.get(name=product_name)
+                product.authorized_users.remove(user)
+                product.save()
```

Comment:
I have 200+ products in my instance. I think this code might become slow in that case, checking every product one by one on every login? Or am I wrong. Could you look at maybe performing one query to get all existing products and prefetch the authorized users so it will be cheap to check if an update is needed?

Reply:
Thanks @valentijnscholten for the suggestions, I'm working on them

Reply:
Now I filter only current user's product in order to be more efficient